libs/langchain-google-common/src/experimental/utils/media_core.ts

```diff
@@ -0,0 +1,504 @@
+    const oldPath = oldBlob.path;
+    const metadata = oldBlob?.metadata ?? {};
+    metadata.langchainOldPath = oldPath;
+    const newBlob = new MediaBlob({
+      ...oldBlob,
+      metadata,
+    });
+    if (newPath) {
+      newBlob.path = newPath;
+    } else if (newBlob.path) {
+      delete newBlob.path;
+    }
+    return newBlob;
+  }
+
+  protected async _validBlobPrefixPath(
+    blob: MediaBlob,
+    opts?: BlobStoreStoreOptions
+  ): Promise<MediaBlob> {
+    const prefix = opts?.pathPrefix ?? "";
+    const suffix = this._blobPathSuffix(blob);
+    const newPath = `${prefix}${suffix}`;
+    return this._newBlob(blob, newPath);
+  }
+
+  protected async _validBlobPrefixUuid(
+    blob: MediaBlob,
+    opts?: BlobStoreStoreOptions
+  ): Promise<MediaBlob> {
+    const prefix = opts?.pathPrefix ?? "";
+    const suffix = uuidv4(); // TODO - option to specify version?
+    const newPath = `${prefix}${suffix}`;
+    return this._newBlob(blob, newPath);
+  }
+
+  protected async _validBlobRemovePath(
+    blob: MediaBlob,
+    _opts?: BlobStoreStoreOptions
+  ): Promise<MediaBlob> {
+    return this._newBlob(blob, undefined);
+  }
+
+  /**
+   * Based on the blob and options, return a blob that has a valid path
+   * that can be saved.
+   * @param blob
+   * @param opts
+   */
+  protected async _validStoreBlob(
+    blob: MediaBlob,
+    opts?: BlobStoreStoreOptions
+  ): Promise<MediaBlob | undefined> {
+    if (await this._hasValidPath(blob, opts)) {
+      return blob;
+    }
+    switch (opts?.actionIfInvalid) {
+      case "ignore":
+        return blob;
+      case "prefixPath":
+        return this._validBlobPrefixPath(blob, opts);
+      case "prefixUuid":
+        return this._validBlobPrefixUuid(blob, opts);
+      case "removePath":
+        return this._validBlobRemovePath(blob, opts);
+      default:
+        return undefined;
+    }
+  }
+
+  async store(
+    blob: MediaBlob,
+    opts: BlobStoreStoreOptions = {}
+  ): Promise<MediaBlob | undefined> {
+    const allOpts: BlobStoreStoreOptions = {
+      ...this.defaultStoreOptions,
+      ...opts,
+    };
+    const validBlob = await this._validStoreBlob(blob, allOpts);
+    if (typeof validBlob !== "undefined") {
+      const validKey = await validBlob.asUri();
+      await this.mset([[validKey, validBlob]]);
+      const savedKey = await validBlob.asUri();
+      return await this.fetch(savedKey);
+    }
+    return undefined;
+  }
+
+  protected async _missingFetchBlobEmpty(
+    path: string,
+    _opts?: BlobStoreFetchOptions
+  ): Promise<MediaBlob> {
+    return new MediaBlob({ path });
+  }
+
+  protected async _missingFetchBlob(
+    path: string,
+    opts?: BlobStoreFetchOptions
+  ): Promise<MediaBlob | undefined> {
+    switch (opts?.actionIfBlobMissing) {
+      case "emptyBlob":
+        return this._missingFetchBlobEmpty(path, opts);
+      default:
+        return undefined;
+    }
+  }
+
+  async fetch(
+    key: string | MediaBlob,
```

Comment:
I think my thinking here was that, eventually, it would make sense for developers to be able to pass MediaBlobs around directly (ie - data that comes from image generation) without having to save it to an intermediary step.

I am not strongly committed to this.